sql/src/main/java/io/crate/operation/scalar/stats/PercentileFunction.java

```diff
@@ -0,0 +1,152 @@
+
+    private PercentileFunction(FunctionInfo info) {
+        this.info = info;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+
+    protected static double[] toDoubleArray(Object[] array) {
+        DoubleArrayList values = new DoubleArrayList(array.length);
+        for (Object value : array) {
+            if (value != null) {
+                values.add(DataTypes.DOUBLE.value(value));
+            }
+        }
+        return values.toArray();
+    }
+
+    protected static boolean nullOrEmptyArgs(Object arrayValue, Object fractionValue) {
+        return arrayValue == null || fractionValue == null || ((Object[]) arrayValue).length == 0;
+    }
+
+    private static class PercentileContArray extends PercentileFunction<Double[], Number[]> {
+
+        public static void register(ScalarFunctionModule module) {
+            for (DataType dataType : ALLOWED_TYPES) {
+                FunctionIdent ident = new FunctionIdent(NAME, ImmutableList.of(DOUBLE_ARRAY_TYPE, new ArrayType(dataType)));
+                module.register(new PercentileContArray(new FunctionInfo(ident, DOUBLE_ARRAY_TYPE)));
+            }
+        }
+
+        PercentileContArray(FunctionInfo info) {
+            super(info);
+        }
+
+        @Override
+        public Double[] evaluate(Input<Number[]>[] args) {
+            assert args.length == 2;
+            Object arrayValue = args[1].value();
+            Object fractionsValue = args[0].value();
+
+            if (nullOrEmptyArgs(arrayValue, fractionsValue)) {
+                return null;
+            }
+
+            return evaluatePercentiles((Object[]) arrayValue, (Object[]) fractionsValue);
+        }
+
+        private Double[] evaluatePercentiles(Object[] values, Object[] fractions) {
+            final Percentile percentile = new Percentile();
+            percentile.setData(toDoubleArray((values)));
+
+            Double[] result = new Double[fractions.length];
+            for (int idx = 0; idx < fractions.length; idx++) {
+                result[idx] = percentile.evaluate(DataTypes.DOUBLE.value(fractions[idx]) * 100.0);
+            }
+            return result;
+        }
+    }
+
+
+    private static class PercentileCont extends PercentileFunction<Double, Number[]> {
+
+        public static void register(ScalarFunctionModule module) {
+            for (DataType dataType : ALLOWED_TYPES) {
+                FunctionIdent ident
+                    = new FunctionIdent(NAME, ImmutableList.<DataType>of(DataTypes.DOUBLE, new ArrayType(dataType)));
+                module.register(new PercentileCont(new FunctionInfo(ident, DataTypes.DOUBLE)));
+            }
+        }
+
+        PercentileCont(FunctionInfo info) {
+            super(info);
+        }
+
+        @Override
+        public Double evaluate(Input<Number[]>[] args) {
+            assert args.length == 2;
+            Object arrayValue = args[1].value();
+            Object fractionValue = args[0].value();
```

Comment:
I think this should be safe to cast to double because the function is only registered for  <double, array<T>>?
Would avoid the DataTypes.DOUBLE.value call further down.